// bot-express/module/messenger/webclient.js
'use strict';

const axios = require("axios")
const crypto = require("crypto")
const debug = require("debug")("bot-express:messenger")
const cache = require("memory-cache")
const secure_compare = require('secure-compare')


const DEFAULT_TOKEN_STORE = "memory-cache"
const REQUIRED_PARAMETERS = ["org_id"]
const CACHE_PREFIX = "be_messenger_webclient_"

const dialogflow = require('dialogflow');
const uuid = require('uuid');


module.exports = class MessengerWebClient {

  constructor(options){
      for (let p of REQUIRED_PARAMETERS){
          if (!options[p]){
              throw new Error(`Required parameter: "${p}" for webclient configuration not set.`);
          }
      }

      this._org_id = options.org_id;
      this.sdk = null; // TBC
  }
  
  validate_signature(req){
    let signature = req.get('org_id');
    let raw_ord_id=this._org_id;
    let raw_body = Buffer.from(JSON.stringify(req.body));
    debug(`raw_body:${raw_body}`);
    // Signature Validation
    let secret_key = crypto.createHmac("sha1", this._org_id).update(this._org_id,"utf8","hex").digest("hex");
    debug(`org_id signature:${signature},  org_id:${raw_ord_id}, secret org_id:${secret_key}`);
    if (secure_compare(secret_key, signature)){
      return true
    }
    return false
  }

  get_secret(){
      return this._org_id;
  }

  async refresh_token(){
    const key = `${CACHE_PREFIX}${this._org_id}_access_token`
    let access_token
    if (this.token_store === "redis"){
        access_token = await this.redis_client.get(key)
    } else {
        access_token = cache.get(key)
    }
    return;
  }

  async multicast(event, to_list, messages){
    return Promise.reject(new Error("This method is not supported."));
  }

  async send(event, to, messages){
    // If this is test, we will not actually issue call out.
    if (["development", "test"].includes(process.env.BOT_EXPRESS_ENV)){
        debug("This is test so we skip the actual call out.");
        return;
    }

    //return this.sdk.pushMessage(to, messages);
    let url = `url`;//`https://${this._endpoint}/${this._api_version}/bot/message/push`;
    let headers = {
        Authorization: `Bearer ${this._access_token}`
    };
    let body = {
        to: to,
        messages: messages
    };

    debug(`Running send..`);
    return this.request({
        method: "post",
        url: url,
        headers: headers,
        data: body
    });
  }

  async reply_to_collect(event, messages){
      return this.reply(event, messages);
  }

  async reply(event, messages){
    // If this is test, we will not actually issue call out.
    if (["development", "test"].includes(process.env.BOT_EXPRESS_ENV)){
      debug("This is test so we skip the actual call out.");
      return;
    }

    //return this.sdk.replyMessage(event.replyToken, messages);

    let url = `https://${this._endpoint}/${this._api_version}/bot/message/reply`;
    let headers = {
        Authorization: `Bearer ${this._access_token}`
    }
    let body = {
        replyToken: event.replyToken,
        messages: messages
    }
    
    debug(`Running reply..`)
    return this.request({
        method: "post",
        url: url,
        headers: headers,
        data: body
    })
  }

  static extract_events(body){
    return [body];
  }

  static identify_event_type(event){
      let event_type;
      if (event.message){
          if (event.message.is_echo){
              event_type = "echo";
          } else {
              event_type = "message";
          }
      } else if (event.delivery){
          event_type = "delivery";
      } else if (event.read){
          event_type = "read";
      } else if (event.postback){
          event_type = "postback";
      } else if (event.optin){
          event_type = "optin";
      } else if (event.referral){
          event_type = "referral";
      } else if (event.account_linking){
          event_type = "account_linking";
      } else if (event.type && event.type == "bot-express:push"){
          event_type = "bot-express:push";
      } else {
          event_type = "unidentified";
      }
      return event_type;
  }

  static extract_beacon_event_type(event){
      let beacon_event_type = false;
      return beacon_event_type;
  }

  static extract_sender_id(event){
      if (event.type === "bot-express:push"){
          return MessengerFacebook.extract_to_id(event);
      }
      return event.sender.id;
  }

  extract_channel_id(event){
      return event.recipient.id
  }

  static extract_to_id(event){
      return event.to[`${event.to.type}Id`];
  }

  static extract_session_id(event){
      return MessengerFacebook.extract_sender_id(event);
  }

  static extract_param_value(event){
      let param_value;
      if (event.message){
          if (event.message.quick_reply){
              // This is Quick Reply
              param_value = event.message.quick_reply.payload;
          } else if (event.message.text){
              // This is Text Message
              param_value = event.message.text;
          } else if (event.message.attachments){
              // This is Attachemnt
              param_value = event.message;
          }
      } else if (event.postback){
          // This is Postback
          param_value = event.postback;
      }
      return param_value;
  }

  static extract_message(event){
      let message;
      if (event.message){
          message = event.message;
      } else if (event.postback){
          message = event.postback;
      }
      return message;
  }

  static extract_message_text(event){
      let message_text;
      if (event.message){
          if (event.message.quick_reply){
              // This is Quick Reply
              message_text = event.message.quick_reply.payload;
          } else if (event.message.text){
              // This is Text Message
              message_text = event.message.text;
          }
      } else if (event.postback){
          // This is Postback
          message_text = event.postback.payload;
      }
      return message_text;
  }

  static extract_postback_payload(event){
      return event.postback.payload;
  }

  static check_supported_event_type(event, flow){
    return true;
  }

  static identify_message_type(message){
      let message_type;
      if (message.text){
          // Type is text.
          message_type = "text";
      } else if (message.attachment){
          if (["image", "audio", "video", "file"].indexOf(message.attachment.type) !== -1){
              // Type is image, audio, video or file.
              message_type = message.attachment.type;
          } else if (message.attachment.type == "template"){
              if (["button", "generic", "list", "open_graph", "receipt", 
              "airline_boardingpass", "airline_checkin", "airline_itinerary",
              "airline_update"].indexOf(message.attachment.payload.template_type) !== -1){
                  message_type = message.attachment.payload.template_type + "_template";
              }
          } else {
              // This is not Facebook format.
              throw new Error("This is not correct webclient format.");
          }
      } else {
          // This is not Facebook format.
          throw new Error("This is not correct webclient format.");
      }
      return message_type;
  }

  static compile_message(message_format, message_type, message){
      return MessengerWebClient[`_compile_message_from_${message_format}_format`](message_type, message);
  }

  static _compile_message_from_line_format(message_type, message){
      // Facebook format has Text, Audio, Image, Video, File, Button Template, Generic Template, List Template, Open Graph Template, Receipt Template, Airline Boarding Ticket Template, Airline Checkin Template, Airline Itinerary Tempalte, Airline Fight Update Template.
      // quick_replies may be included in any Content-Type.
      // buttons may be included in Templates.

      // ### Threshold for quick_replies
      // -> elements of quick_replies has to be up to 11.
      // -> title in each element has to be up to 20 chars.
      // -> payload in each element has to be 1000 chars.

      // ### Threshold for Text ###
      // -> text has to be up to 640 chars.

      // ### Threshold for Button Template ###
      // -> text has to be up to 640 chars.
      // -> buttons has to be up to 3.
      // -> each button must follow button threshold.

      // ### Threshold for Generic Template ###
      // -> elements has to be up to 10.
      // -> title in each elements has to be up to 80 chars.
      // -> subtitle in each elements has to be up to 80 chars.
      // -> buttons in each elements has to be up to 3.
      // -> each button must follow button threshold.

      // ### Threshold for List Template ###
      // -> elements has to be from 2 to 4.
      // -> global button has to be up to 1.
      // -> title in each elements has to be up to 80 chars.
      // -> subtitle in each elements has to be up to 80 chars.
      // -> button in each elements has to be up to 1.
      // -> each button must follow button threshold.

      // ### Compile Rule Overview
      // -> text: to text
      // -> image: to image
      // -> video: to video
      // -> audio: to audio
      // -> file: to unsupported text
      // -> location: to location *NEED TEST
      // -> sticker: to unsupported text
      // -> imagemap: to unsupported text
      // -> buttons template: to text(w quick reply) or button tempalte
      // -> confirm template: to text(w quick reply) or button template
      // -> carousel template: to generic template

      let compiled_message;

      switch(message_type){
          case "text": {// -> to text
              compiled_message = {
                  text: message.text
              }
              break;
          }
          case "image": {// -> to image
              compiled_message = {
                  attachment: {
                      type: "image",
                      payload: {
                          url: message.originalContentUrl
                      }
                  }
              }
              break;
          }
          case "video": {// -> to video
              compiled_message = {
                  attachment: {
                      type: "video",
                      payload: {
                          url: message.originalContentUrl
                      }
                  }
              }
              break;
          }
          case "audio": {// -> to audio
              compiled_message = {
                  attachment: {
                      type: "audio",
                      payload: {
                          url: message.originalContentUrl
                      }
                  }
              }
              break;
          }
          case "file": {// -> unsupported text
              debug(`*Message type is LINE's ${message_type} and it is not supported in Facebook.`);
              compiled_message = {
                 type: "text",
                 text: `*Message type is LINE's ${message_type} and it is not supported in Facebook.`
              }
              break;
          }
          case "location": {// to location *NEED TEST
              compiled_message = {
                  attachment: {
                      type: "location",
                      title: message.title,
                      payload: {
                          coordinates: {
                              lat: message.latitude,
                              long: message.longitude
                          }
                      }

                  }
              }
              break;
          }
          case "sticker": {// -> to unsupported text
              debug(`*Message type is LINE's ${message_type} and it is not supported in Facebook.`);
              compiled_message = {
                 type: "text",
                 text: `*Message type is LINE's ${message_type} and it is not supported in Facebook.`
              }
              break;
          }
          case "imagemap": {// -> to unsupported text
              debug(`*Message type is LINE's ${message_type} and it is not supported in Facebook.`);
              compiled_message = {
                 type: "text",
                 text: `*Message type is LINE's ${message_type} and it is not supported in Facebook.`
              }
              break;
          }
          case "buttons_template": {// -> to text(w quick reply) or button tempalte
              let uri_included = false;
              let datetimepicker_included = false;

              for (let action of message.template.actions){
                  if (action.type == "uri"){
                      uri_included = true;
                  }
                  if (action.type == "datetimepicker"){
                      datetimepicker_included = true;
                  }
              }

              if (uri_included){
                  // This template message include uri button so we use template message in facebook as well.
                  if (message.template.actions.length > 3){
                      // Not supported since facebook does not allow template message including more than 3 buttons. The threshold of action of line template button is 4.
                      debug(`Compiling template message including more than 3 buttons including uri button from line format to facebook format is not supported. So we compile it to text message.`);
                      compiled_message = {
                          text: message.altText + " *Compiling template message including more than 3 buttons including uri button from line format to facebook format is not supported. So we compile it to text message."
                      }
                      break;
                  }
                  compiled_message = {
                      attachment: {
                          type: "template",
                          payload: {
                              template_type: "button",
                              text: message.template.text,
                              buttons: []
                          }
                      }
                  }
                  for (let action of message.template.actions){
                      if (action.type == "uri"){
                          compiled_message.attachment.payload.buttons.push({
                              type: "web_url",
                              url: action.uri,
                              title: action.label
                          });
                      } else if (action.type == "postback"){
                          compiled_message.attachment.payload.buttons.push({
                              type: "postback",
                              title: action.label,
                              payload: action.data
                          });
                      } else if (action.type == "message"){
                          compiled_message.attachment.payload.buttons.push({
                              type: "postback",
                              title: action.label,
                              payload: action.text
                          });
                      }
                      // We remove datetimepicker button since its not supported in facebook.
                  }
              } else {
                  // This template message does not include uri.
                  // If the number of button is just 1 and it is datetimepicker, we create plane text message.
                  // Otherwise, it can be postback or message so we use quick reply.

                  if (datetimepicker_included && message.template.actions.length == 1){
                      // Message has datetimepicker which is not supported in facebook.
                      // We compile this message to plane text.
                      compiled_message = {
                          text: message.altText
                      }
                  } else {
                      compiled_message = {
                          text: message.template.text,
                          quick_replies: []
                      }
                      for (let action of message.template.actions){
                          if (action.type == "postback"){
                              compiled_message.quick_replies.push({
                                  content_type: "text",
                                  title: action.label,
                                  payload: action.data
                              });
                          } else if (action.type == "message"){
                              compiled_message.quick_replies.push({
                                  content_type: "text",
                                  title: action.label,
                                  payload: action.text
                              });
                          }
                      }
                  }
              }
              break;
          }
          case "confirm_template": {// -> to text(w quick reply) or button tempalte
              let uri_included = false;
              let datetimepicker_included = false;

              for (let action of message.template.actions){
                  if (action.type == "uri"){
                      uri_included = true;
                  }
                  if (action.type == "datetimepicker"){
                      datetimepicker_included = true;
                  }
              }

              if (uri_included){
                  // This template message include uri button so we use template message in facebook as well.
                  if (message.template.actions.length > 3){
                      // Not supported since facebook does not allow template message including more than 3 buttons. The threshold of action of line template button is 4.
                      debug(`Compiling template message including more than 3 buttons including uri button from line format to facebook format is not supported. So we compile it to text message.`);
                      compiled_message = {
                          text: message.altText + " *Compiling template message including more than 3 buttons including uri button from line format to facebook format is not supported. So we compile it to text message."
                      }
                      break;
                  }
                  compiled_message = {
                      attachment: {
                          type: "template",
                          payload: {
                              template_type: "button",
                              text: message.template.text,
                              buttons: []
                          }
                      }
                  }
                  for (let action of message.template.actions){
                      if (action.type == "uri"){
                          compiled_message.attachment.payload.buttons.push({
                              type: "web_url",
                              url: action.uri,
                              title: action.label
                          });
                      } else {
                          compiled_message.attachment.payload.buttons.push({
                              type: "postback",
                              title: action.label,
                              payload: action.data
                          });
                      }

                  }
              } else {
                  // This template message does not include uri.
                  // If the number of button is just 1 and it is datetimepicker, we create plane text message.
                  // Otherwise, it can be postback or message so we use quick reply.

                  if (datetimepicker_included && message.template.actions.length == 1){
                      // Message has datetimepicker which is not supported in facebook.
                      // We compile this message to plane text.
                      compiled_message = {
                          text: message.altText
                      }
                  } else {
                      compiled_message = {
                          text: message.template.text,
                          quick_replies: []
                      }
                      for (let action of message.template.actions){
                          if (action.type == "postback"){
                              compiled_message.quick_replies.push({
                                  content_type: "text",
                                  title: action.label,
                                  payload: action.data
                              });
                          } else if (action.type == "message"){
                              compiled_message.quick_replies.push({
                                  content_type: "text",
                                  title: action.label,
                                  payload: action.text
                              });
                          }
                      }
                  }
              }
              break;
          }
          case "carousel_template": {// -> generic template
              compiled_message = {
                  attachment: {
                      type: "template",
                      payload: {
                          template_type: "generic",
                          elements: []
                      }
                  }
              }
              for (let column of message.template.columns){
                  let element = {
                      title: column.text,
                      image_url: column.thumbnailImageUrl,
                      buttons: []
                  }

                  let uri_included = false;
                  let datetimepicker_included = false;

                  for (let action of column.actions){
                      if (action.type == "uri"){
                          uri_included = true;
                      }
                      if (action.type == "datetimepicker"){
                          datetimepicker_included = true;
                      }
                  }

                  if (uri_included){
                      if (column.actions.length > 3){
                          // Not supported since facebook does not allow template message including more than 3 buttons. line's threshold is 3, too.
                          debug(`Compiling template messege including more than 3 buttons including uri button from line format to facebook format is not supported.`);
                          compiled_message = {
                              text: message.altText + " *Compiling template messege including more than 3 buttons including uri button from line format to facebook format is not supported."
                          }
                          break;
                      }
                  }
                  if (datetimepicker_included && column.actions.length == 1){
                      debug(`Compiling template message including just 1 button which is datetimepicker is not supported.`);
                      compiled_message = {
                          text: message.altText + " *Compiling template message including just 1 button which is datetimepicker is not supported."
                      }
                      break;
                  }
                  for (let action of column.actions){
                      if (action.type == "postback"){
                          element.buttons.push({
                              type: "postback",
                              title: action.label,
                              payload: action.data
                          });
                      } else if (action.type == "message"){
                          element.buttons.push({
                              type: "postback",
                              title: action.label,
                              payload: action.text
                          });
                      } else if (action.type == "uri"){
                          element.buttons.push({
                              type: "web_url",
                              url: action.uri,
                              title: action.label
                          });
                      }
                      // We remove datetimepicker since it's not supported in facebook.
                  }
                  compiled_message.attachment.payload.elements.push(element);
              }
              break;
          }
          default: {
              debug(`Message type is LINE's ${message_type} and it is not supported in Facebook.`);
              compiled_message = {
                 type: "text",
                 text: `*Message type is LINE's ${message_type} and it is not supported in Facebook.`
              }
              break;
          }
      }
      return compiled_message
  }

  /**
  @deprecated
  */
  static translate_message(translater, message_type, message, sender_language){
    throw new Error("This method translate_message is not supported.");
  }

  async request(options){
      let response = await axios.request(options).catch(e => {
          let error_message = `Failed.`
          if (e.resopnse){
              error_message += ` Status code: ${e.response.status}. Payload: ${JSON.stringify(e.response.data)}`;
          }
          throw Error(error_message)
      })
      return response.data
  }
}
